parlai/utils/torch.py

```diff
@@ -235,3 +238,229 @@ def forward(self, xs):
         Identity.
         """
         return xs
+
+
+Chunk = TypeVar('Chunk')
+
+
+PipelineWorkItem = namedtuple(
+    'PipelineWorkItem', ['chunk_idx', 'layer_nos', 'next_device']
+)
+
+
+class PipelineHelper(object):
+    """
+    PipelineHelper assists with implementing pipelining in model parallelism.
+
+    For a tutorial on model parallelism, as it's implemented in parts of ParlAI,
+    see https://pytorch.org/tutorials/intermediate/model_parallel_tutorial.html.
+    """
+
+    @staticmethod
+    def guess_split_size(item: Chunk, num_gpus: Optional[int] = None, dim=0) -> int:
+        """
+        Estimate the number of chunks we should split the batch into.
+
+        Uses some silly heuristics.
+        """
+        if num_gpus is None:
+            num_gpus = torch.cuda.device_count()  # type: ignore
+        if isinstance(item, torch.Tensor):
+            return max(1, item.size(dim) // int(num_gpus * 2))
```

Comment:
why do we divide by `num_gpus * 2` ??

Reply:
That's the heuristic. :) I've found it by just toying. Really it should be tuned per-model, but I'd rather have something that works roughly fine out of the box for all models.

![image](https://user-images.githubusercontent.com/31896/76988603-5471d080-691b-11ea-8cc9-b851e5b545f0.png)

The gist is that we want to minimize the size of this bubble, while not running out of memory. If you use the same number of chunks as GPUS, it's fairly straightforward to see that you will have a large bubble. If you chunk via 2x the number of GPUs, then the bubble becomes smaller, but I stretch the time dimension out a lot. So there's sorta a fine balance. In @mrshenli's excellent tutorial, he shows this graph, but it's specific to a single model:

![image](https://user-images.githubusercontent.com/31896/76988819-be8a7580-691b-11ea-8f8e-a018bd6fa691.png)

Reply:
nice, thanks for detailed explanation!